builder-jql/src/main/java/hu/blackbelt/judo/meta/expression/builder/jql/function/temporal/ConstructorTransformer.java

```diff
@@ -94,35 +67,114 @@ public Expression apply(Expression argument, JqlFunction functionCall, Expressio
         }
     }
 
+    private DateConstructionExpression getDateConstructionExpression(JqlFunction functionCall, ExpressionBuildingVariableResolver context) {
+        EList<FunctionParameter> parameters = functionCall.getParameters();
+
+        JqlTransformerUtils.validateParameterCount(functionCall.getName(), parameters, 3);
+
+        return newDateConstructionExpressionBuilder()
+                .withYear((IntegerExpression) jqlTransformers.transform(parameters.get(0).getExpression(), context))
+                .withMonth((IntegerExpression) jqlTransformers.transform(parameters.get(1).getExpression(), context))
+                .withDay((IntegerExpression) jqlTransformers.transform(parameters.get(2).getExpression(), context))
+                .build();
+    }
+
+    private TimeConstructionExpression getTimeConstructionExpression(JqlFunction functionCall, ExpressionBuildingVariableResolver context) {
+        EList<FunctionParameter> parameters = functionCall.getParameters();
+
+        int parameterCount = JqlTransformerUtils.validateParameterCount(functionCall.getName(), parameters, 2, 3, 4);
+
+        final IntegerExpression second;
+        if (parameterCount > 2) {
+            second = (IntegerExpression) jqlTransformers.transform(parameters.get(2).getExpression(), context);
+        } else {
+            second = IntegerConstantBuilder.create().withValue(BigInteger.ZERO).build();
+        }
+
+        final IntegerExpression millisecond;
+        if (parameterCount == 4) {
+            millisecond = (IntegerExpression) jqlTransformers.transform(parameters.get(3).getExpression(), context);
+        } else {
+            millisecond = IntegerConstantBuilder.create().withValue(BigInteger.ZERO).build();
+        }
+
+        return newTimeConstructionExpressionBuilder()
+                .withHour((IntegerExpression) jqlTransformers.transform(parameters.get(0).getExpression(), context))
+                .withMinute((IntegerExpression) jqlTransformers.transform(parameters.get(1).getExpression(), context))
+                .withSecond(second)
+                .withMillisecond(millisecond)
+                .build();
+    }
+
+    private TimestampConstructionExpression getTimestampConstructionExpression(JqlFunction functionCall, ExpressionBuildingVariableResolver context) {
+        int parameterCount = JqlTransformerUtils.validateParameterCount(functionCall.getName(), functionCall.getParameters(), 1, 2, 5, 6, 7);
```

Comment:
Magic number, I have no idea what these numbers represent. We should introduce named constants.

Reply:
A timestamp construction to be valid, 1, 2, 5, 6 or 7 parameters are accepted.
If there are 1 or 2 parameters, 1st must be a date typed argument, 2nd is time typed argument.
For 5, 6 and 7 => year, month, day, hour, minute, second and millisecond all of them integer typed arguments.
This is due to the lack of named parameters.

I can add an inline comment before this check to improve understandability of the code.